### PR TITLE
Task 6

### PR DESCRIPTION
**Pull request by denja2**

### CODE REVIEW

File: ConsoleApp/BL/Services/WeatherService.cs [hunk 1]
```diff
@@ -26,15 +29,26 @@ public async Task<string> GetWeatherAsync(string cityName)
 
             weather = SetWeatherDescription(weather);
 
-            return $"In {weather.name} {weather.main.temp}°C now. {weather.weather.First().description}\n";
+            return $"\nIn {weather.name} {weather.main.temp}°C now. {weather.weather.First().description}\n";
+        }
+
+        public async Task<List<WeatherForecastDTO>> GetWeatherForecastAsync(string cityName, int days)
+        {
+            _validator.ValidateCityName(cityName);
```
Comment (denja2) [inline ConsoleApp/BL/Services/WeatherService.cs:37]: Validate request model

File: ConsoleApp/BL/Validators/Validator.cs [hunk 2]
```diff
@@ -3,13 +3,21 @@
...
             if (string.IsNullOrWhiteSpace(cityName))
             {
-                throw new ValidatorException("\nIncorrectly entered data");
+                throw new ValidatorException("\nInvalid data entered");
+            }
+        }
+
+        public void ValidateNumberOfDays(int days)
+        {
+            if (days > 7 || days <= 0)
```
Comment (denja2) [inline ConsoleApp/BL/Validators/Validator.cs:18]: From config

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs [hunk 3]
```diff
@@ -40,5 +71,24 @@ public async Task<Root> GetWeatherAsync(string cityName)
                 return null;
             }
         }
+
+        private async Task<Geolocation> GetWeatherCoord(string cityName)
```
Comment (denja2) [inline ConsoleApp/DAL/Repositories/WeatherRepository.cs:75]: Async

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs [hunk 4]
```diff
@@ -40,5 +71,24 @@ public async Task<Root> GetWeatherAsync(string cityName)
                 return null;
```
Comment (denja2) [inline ConsoleApp/DAL/Repositories/WeatherRepository.cs:71]: Remove

File: ConsoleApp/BL/Services/WeatherService.cs (hunk 1 above)
Comment (Xodoc) [reply ConsoleApp/BL/Services/WeatherService.cs:37]: done

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs (hunk 3 above)
Comment (Xodoc) [reply ConsoleApp/DAL/Repositories/WeatherRepository.cs:75]: done

File: ConsoleApp/BL/Validators/Validator.cs (hunk 2 above)
Comment (Xodoc) [reply ConsoleApp/BL/Validators/Validator.cs:18]: Done

File: ConsoleApp/BL/Services/WeatherService.cs [hunk 5]
```diff
@@ -26,31 +29,95 @@ public async Task<string> GetWeatherAsync(string cityName)
...
+
+        public async Task<string> GetWeatherForecastAsync(string cityName, int days)
+        {
+            _validator.ValidateModel(cityName, days);
+
+            var weatherForecast = await _weatherRepository.GetWeatherForecastAsync(cityName, days);
+
+            var weatherForecastDtos = MapToWeatherForecastDTOs(weatherForecast);
+
+            weatherForecastDtos = SetWeatherForecastDescription(weatherForecastDtos);
```
Comment (denja2) [inline ConsoleApp/BL/Services/WeatherService.cs:43]: To mapper

File: ConsoleApp/BL/Services/WeatherService.cs [hunk 6]
```diff
@@ -26,31 +29,95 @@ public async Task<string> GetWeatherAsync(string cityName)
...
+
+            var responseMessage = "";
+
+            foreach (var item in weatherForecastDtos)
```
Comment (denja2) [inline ConsoleApp/BL/Services/WeatherService.cs:47]: string.Join

File: ConsoleApp/BL/Services/WeatherService.cs [hunk 7]
```diff
@@ -26,31 +29,95 @@ public async Task<string> GetWeatherAsync(string cityName)
...
+                if (weatherForecastDto[i].Temp < 0)
+                    weatherForecastDto[i].Description = "Dress warmly.";
+
+                if (weatherForecastDto[i].Temp >= 0 && weatherForecastDto[i].Temp <= 20)
+                    weatherForecastDto[i].Description = "It's fresh.";
+
+                if (weatherForecastDto[i].Temp >= 20 && weatherForecastDto[i].Temp <= 30)
+                    weatherForecastDto[i].Description = "Good weather.";
+
+                if (weatherForecastDto[i].Temp >= 30)
```
Comment (denja2) [inline ConsoleApp/BL/Services/WeatherService.cs:94]: Move to separate method

File: ConsoleApp/BL/Validators/Validator.cs [hunk 8]
```diff
@@ -1,15 +1,35 @@
...
             if (string.IsNullOrWhiteSpace(cityName))
             {
-                throw new ValidatorException("\nIncorrectly entered data");
+                throw new ValidatorException("\nInvalid data entered");
+            }
+        }
+
+        public void ValidateModel(string cityName, int days)
+        {
+            if (string.IsNullOrWhiteSpace(cityName))
```
Comment (denja2) [inline ConsoleApp/BL/Validators/Validator.cs:26]: use ValidateCityName

File: ConsoleApp/ConsoleApp/Commands/GetWeatherCommand.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+using System;
+using System.Threading.Tasks;
+
+namespace ConsoleApp.Commands
+{
+    public class GetWeatherCommand : ICommand
+    {
+        private readonly IWeatherService _weatherService;
+
+        public string Title => "\n0) Show weather";
```
Comment (denja2) [inline ConsoleApp/ConsoleApp/Commands/GetWeatherCommand.cs:12]: set number in main method

File: ConsoleApp/ConsoleApp/Commands/GetWeatherForecastCommand.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+                Console.Write("\nInput number of days: ");
+                var days = int.Parse(Console.ReadLine());
+                Console.Clear();
+
+                var weatherForecast = await _weatherService.GetWeatherForecastAsync(cityName, days);
+
+                Console.WriteLine(weatherForecast);
+
+                Console.WriteLine("\nPress any key to continue...");
```
Comment (denja2) [inline ConsoleApp/ConsoleApp/Commands/GetWeatherForecastCommand.cs:34]: move to main

File: ConsoleApp/ConsoleApp/Program.cs [hunk 11]
```diff
@@ -14,44 +17,53 @@ public class Program
     {
         private static IWeatherService _weatherService;
 
-        private static async Task ShowWeather()
+        private static async Task Menu()
         {
-            try
+            ICommand getWeather = new GetWeatherCommand(_weatherService);
```
Comment (denja2) [inline ConsoleApp/ConsoleApp/Program.cs:22]: var

File: ConsoleApp/IntegrationTests/Services/WeatherServiceIntegrationTest.cs [hunk 12]
```diff
@@ -54,13 +53,33 @@ public async void GetWeatherAsync_WhenSendingCorrectCityName_GettingWeatherMessa
...
             //Assert
             Assert.Equal(expectedMessage, actualResult.Message);
         }
+
+        [Theory]
+        [InlineData("Minsk", 3)]
+        public async void GetWeatherForecastAsync_WhenSendingCorrectData_GettingWeatherForecastMessage(string cityName, int days) 
+        {
+            //Arrange
+            var des = new string[]
```
Comment (denja2) [inline ConsoleApp/IntegrationTests/Services/WeatherServiceIntegrationTest.cs:70]: self descriptive name

File: ConsoleApp/IntegrationTests/Services/WeatherServiceIntegrationTest.cs [hunk 13]
```diff
@@ -54,13 +53,33 @@ public async void GetWeatherAsync_WhenSendingCorrectCityName_GettingWeatherMessa
...
+                "It's fresh\\.",
+                "Good weather\\.",
+                "It's time to go to the beach\\."
+            };
+
+            //Act
+            var actualResult = await _weatherService.GetWeatherForecastAsync(cityName, days);
+
+            //Assert
+            Assert.Matches($"^In {cityName} ([0-9])(.*?)\\.|(\\B\\W)(.*?)\\. ({des[0]}|{des[1]}|{des[2]}|{des[3]})$", actualResult);
```
Comment (denja2) [inline ConsoleApp/IntegrationTests/Services/WeatherServiceIntegrationTest.cs:82]: Check this

File: ConsoleApp/Tests/Services/WeatherServiceTest.cs [hunk 14]
```diff
@@ -65,5 +64,49 @@ public async void GetWeatherAsync_WhenSendingIncorrectCityName_GettingWeatherMes
...
+        {
+            //Arrange
+            var weatherForecast = _fixture.Create<WeatherForecast>();
+            var expectedMessage = "";
+
+            foreach (var day in weatherForecast.Daily)
+            {
+                day.Temp.day = temp;
+                day.Weather[0].description = description;
+                expectedMessage += $"{weatherForecast.CityName} weather forecast: {day.Temp.day}°C. {day.Weather[0].description}\n";
```
Comment (denja2) [inline ConsoleApp/Tests/Services/WeatherServiceTest.cs:82]: ???

File: ConsoleApp/BL/Services/WeatherService.cs (hunk 6 above)
Comment (Xodoc) [reply ConsoleApp/BL/Services/WeatherService.cs:47]: done

File: ConsoleApp/ConsoleApp/Commands/GetWeatherCommand.cs (hunk 9 above)
Comment (Xodoc) [reply ConsoleApp/ConsoleApp/Commands/GetWeatherCommand.cs:12]: done

File: ConsoleApp/ConsoleApp/Program.cs (hunk 11 above)
Comment (Xodoc) [reply ConsoleApp/ConsoleApp/Program.cs:22]: done

File: ConsoleApp/IntegrationTests/Services/WeatherServiceIntegrationTest.cs (hunk 13 above)
Comment (Xodoc) [reply ConsoleApp/IntegrationTests/Services/WeatherServiceIntegrationTest.cs:82]: done

File: ConsoleApp/IntegrationTests/Services/WeatherServiceIntegrationTest.cs (hunk 12 above)
Comment (Xodoc) [reply ConsoleApp/IntegrationTests/Services/WeatherServiceIntegrationTest.cs:70]: done

File: ConsoleApp/BL/Validators/Validator.cs (hunk 8 above)
Comment (Xodoc) [reply ConsoleApp/BL/Validators/Validator.cs:26]: done

File: ConsoleApp/BL/Services/WeatherService.cs (hunk 5 above)
Comment (Xodoc) [reply ConsoleApp/BL/Services/WeatherService.cs:43]: done

File: ConsoleApp/BL/Services/WeatherService.cs (hunk 7 above)
Comment (Xodoc) [reply ConsoleApp/BL/Services/WeatherService.cs:94]: done

File: ConsoleApp/Tests/Services/WeatherServiceTest.cs (hunk 14 above)
Comment (Xodoc) [reply ConsoleApp/Tests/Services/WeatherServiceTest.cs]: done

**Review by alesyalis — APPROVED** [review]

File: ConsoleApp/ConsoleApp/Program.cs [hunk 15]
```diff
@@ -14,44 +17,55 @@ public class Program
...
+                    if (input == 2)
+                    {
+                        flag = false; return;
+                    }
+                    else if (input >= 0 && input <= 1)
+                    {
+                        await commands[input].Execute();
+                        Console.WriteLine("\nPress any key to continue...");
+                        Console.ReadKey();
+                    }
```
Comment (denja2) [inline ConsoleApp/ConsoleApp/Program.cs:56]: remove

Comment (Xodoc) [reply ConsoleApp/ConsoleApp/Program.cs:56]: done

**Pull request merged**